### PR TITLE
Support `upper` and `lower` in `strings_udf`

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR adds support for the following two functions in `strings_udf`:

- `str.upper()`
- `str.lower()`

Part of https://github.com/rapidsai/cudf/issues/9639

### CODE REVIEW

File: python/strings_udf/cpp/src/strings/udf/shim.cu [hunk 1]
```diff
@@ -227,3 +229,72 @@ extern "C" __device__ int udf_string_from_string_view(int* nb_retbal,
...
+
+  cudf::strings::udf::chars_tables tables;
+  tables.flags_table                = flags_table_ptr;
+  tables.cases_table                = cases_table_ptr;
+  tables.special_case_mapping_table = special_table_ptr;
+
+  *udf_str_ptr = to_upper(tables, *st_ptr);
+
+  return 0;
+}
```
Comment (davidwendt) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:300]: Is the `lower()` definition missing here?

Comment (brandon-b-miller) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:300]: Added 👍

File: python/cudf/cudf/tests/test_udf_masked_ops.py [hunk 2]
```diff
@@ -876,7 +876,6 @@ def func(row):
     run_masked_udf_test(func, str_udf_data, check_dtype=False)
 
 
-@string_udf_test
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_udf_masked_ops.py:879]: Was this removal needed? or a typo?

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/tests/test_udf_masked_ops.py [hunk 3]
```diff
@@ -876,7 +876,6 @@ def func(row):
...
 @pytest.mark.parametrize("strip_char", ["1", "a", "12", " ", "", ".", "@"])
```
Comment (galipremsagar) [inline python/cudf/cudf/tests/test_udf_masked_ops.py:879]:
```suggestion
@string_udf_test
@pytest.mark.parametrize("strip_char", ["1", "a", "12", " ", "", ".", "@"])
```

**Review by galipremsagar — CHANGES_REQUESTED** [review]
Looks like the CI failures are around this change.

File: python/cudf/cudf/tests/test_udf_masked_ops.py (hunk 2 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_udf_masked_ops.py]: 👍

File: python/strings_udf/strings_udf/_lib/tables.pyx [hunk 4]
```diff
@@ -1,14 +1,26 @@
...
+
+
+def get_character_cases_table_ptr():
+    cdef const uint16_t* tbl_ptr = cpp_get_character_cases_table()
+    return np.uint64(<uintptr_t>tbl_ptr)
+
+
+def get_special_case_mapping_table_ptr():
+    cdef const void* tbl_ptr = cpp_get_special_case_mapping_table()
+    return np.uint64(<uintptr_t>tbl_ptr)
```
Comment (wence-) [inline python/strings_udf/strings_udf/_lib/tables.pyx:26]: You want [`numpy.uintp`](https://numpy.org/doc/stable/reference/arrays.scalars.html#numpy.uintp) not `uint64`

File: python/strings_udf/cpp/src/strings/udf/shim.cu [hunk 5]
```diff
@@ -270,3 +271,57 @@ extern "C" __device__ int rstrip(int* nb_retval,
 
   return 0;
 }
+
+extern "C" __device__ int upper(int* nb_retval,
+                                void* udf_str,
+                                void* const* st,
+                                std::int64_t flags_table,
+                                std::int64_t cases_table,
+                                std::int64_t special_table)
```
Comment (wence-) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:280]: These are pointers being represented by integers I think. So you want `std::uintptr_t` not `int64_t` (although it is only optional in C++11 and later, I think it is provided on all sane platforms).

This comment applies throughout (and indeed in existing code where you accept pointers as integers and then call `reinterpret_cast<some_ptr*>(some_integer)`)

Comment (davidwendt) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:280]: Just curious why these are not `void*` as well?

File: python/strings_udf/cpp/src/strings/udf/shim.cu [hunk 6]
```diff
@@ -270,3 +271,57 @@ extern "C" __device__ int rstrip(int* nb_retval,
...
+  tables.special_case_mapping_table = special_table_ptr;
+
+  *udf_str_ptr = to_upper(tables, *st_ptr);
+
+  return 0;
+}
+
+extern "C" __device__ int lower(int* nb_retval,
+                                void* udf_str,
+                                void* const* st,
```
Comment (wence-) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:304]: Does this need to be a `void **` or can it just be a `void *` (I note that inside you cast to `string_view *` and then dereference, so I think you can strip a `*` everywhere).

File: python/strings_udf/cpp/src/strings/udf/udf_apis.cu [hunk 7]
```diff
@@ -42,7 +42,8 @@ namespace {
 struct udf_string_to_string_view_transform_fn {
   __device__ cudf::string_view operator()(cudf::strings::udf::udf_string const& dstr)
   {
-    return cudf::string_view{dstr.data(), dstr.size_bytes()};
+    return dstr.data() == nullptr ? cudf::string_view{"", 0}
+                                  : cudf::string_view{dstr.data(), dstr.size_bytes()};
```
Comment (wence-) [inline python/strings_udf/cpp/src/strings/udf/udf_apis.cu:46]:
```suggestion
    return dstr.data() == nullptr ? cudf::string_view{}
                                  : cudf::string_view{dstr.data(), dstr.size_bytes()};
```

I think `string_view` has a default empty constructor.

**Review by galipremsagar — APPROVED** [review]

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 5 above)
Comment (brandon-b-miller) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:280]: Let me make sure all is right here. Each of the functions that return a pointer to an input mapping table returns its own type of pointer. For the flags table its `uint8_t*`, for the cases table its `uint16_t`. For the special case mapping table it's a `special_case_mapping*`. Would the correct thing to do in this case be to receive each of these as a `uintptr_t` and then carry them through the python into the lowering as a `np.uintp`? Then these shim functions could accept a `uintptr_t` here.

File: python/strings_udf/strings_udf/lowering.py [hunk 8]
```diff
@@ -324,6 +345,74 @@ def id_func_impl(context, builder, sig, args):
...
+        @cuda_lower(id_func, string_view)
+        def id_func_impl(context, builder, sig, args):
+            str_ptr = builder.alloca(args[0].type)
+            builder.store(args[0], str_ptr)
+
+            # Lookup table required for conversion functions
+            # must be resolved at runtime after context initialization,
+            # therefore cannot be a global variable
+            flags_tbl_ptr = context.get_constant(
+                types.int64, character_flags_table_ptr
```
Comment (wence-) [inline python/strings_udf/strings_udf/lowering.py:365]: Probably `types.uintp` throughout

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 5 above)
Comment (davidwendt) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:280]: I think it would be more correct these to be pointer types than `int64_t` types.

File: python/strings_udf/cpp/src/strings/udf/shim.cu [hunk 9]
```diff
@@ -270,3 +271,57 @@ extern "C" __device__ int rstrip(int* nb_retval,
...
+    reinterpret_cast<cudf::strings::detail::character_flags_table_type*>(flags_table);
+  auto cases_table_ptr =
+    reinterpret_cast<cudf::strings::detail::character_cases_table_type*>(cases_table);
+  auto special_table_ptr =
+    reinterpret_cast<cudf::strings::detail::special_case_mapping*>(special_table);
+
+  cudf::strings::udf::chars_tables tables;
+  tables.flags_table                = flags_table_ptr;
+  tables.cases_table                = cases_table_ptr;
+  tables.special_case_mapping_table = special_table_ptr;
```
Comment (wence-) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:322]: Can we not do: `cudf::strings::udf::chars_tables tables{flags_table_ptr, cases_table_ptr, special_table_ptr};` ?

File: python/strings_udf/cpp/src/strings/udf/shim.cu [hunk 10]
```diff
@@ -270,3 +271,57 @@ extern "C" __device__ int rstrip(int* nb_retval,
...
 }
+
+extern "C" __device__ int upper(int* nb_retval,
+                                void* udf_str,
+                                void* const* st,
+                                std::int64_t flags_table,
+                                std::int64_t cases_table,
+                                std::int64_t special_table)
+{
+  auto udf_str_ptr = new (udf_str) udf_string;
```
Comment (wence-) [inline python/strings_udf/cpp/src/strings/udf/shim.cu:282]: This calls the default constructor, but this object is never read anywhere, and is then overwritten on line 297 (`to_upper` returns a `udf_string`). So what is going on here?

Does this want to disappear and line 297 be `*reinterpret_cast<udf_string *>(udf_str) = to_upper(tables, *st_ptr);` ?

**Review by wence- — CHANGES_REQUESTED** [review]
A bunch of nits around cross-calling ABI and type-punning pointers and integers.

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 5 above)
Comment (wence-) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:280]: If you can interact with ctypes I think you can carry around a `ctypes.cvoidp`.

File: python/strings_udf/cpp/src/strings/udf/udf_apis.cu (hunk 7 above)
Comment (davidwendt) [reply python/strings_udf/cpp/src/strings/udf/udf_apis.cu:46]: > I think `string_view` has a default empty constructor.

You are correct.

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 10 above)
Comment (davidwendt) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:282]: No the object must be initialized by a constructor in order for it to be valid. Otherwise the `operator=` will crash if the target has garbage in it.

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 6 above)
Comment (brandon-b-miller) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:301]: I am not sure I follow here. This follows the pattern from the rest of the shim functions when a `string_view` is an arg. IIUC `st` is pointing directly to the struct, so only one level of pointing right?

File: python/strings_udf/strings_udf/lowering.py [hunk 11]
```diff
@@ -324,6 +345,74 @@ def id_func_impl(context, builder, sig, args):
     return deco
 
 
+def create_upper_or_lower(id_func):
+    """
+    Provide a wrapper around numba's low-level extension API which
+    produces the boilerplate needed to implement a unary function
+    of a string.
```
Comment (wence-) [inline python/strings_udf/strings_udf/lowering.py:352]: Docstring wants changed?

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 10 above)
Comment (wence-) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:282]: Thanks!

File: python/strings_udf/cpp/src/strings/udf/shim.cu (hunk 6 above)
Comment (wence-) [reply python/strings_udf/cpp/src/strings/udf/shim.cu:301]: Most functions here (e.g. https://github.com/rapidsai/cudf/blob/418cbc69b336f34b06e1de21ae2c016b40f1e6fc/python/strings_udf/cpp/src/strings/udf/shim.cu#L186-L193) that take a point to `string_view` take the argument as `void const *`, not `void* const *`. Why is this different?

**Comment by brandon-b-miller** [comment]
rerun tests

**Review by wence- — APPROVED** [review]
Thanks @brandon-b-miller

**Review by davidwendt — APPROVED** [review]
Approving C++ code.

**Comment by brandon-b-miller** [comment]
some test failures here that seem to be related to a previous PR that I am looking into now.

**Comment by brandon-b-miller** [comment]
I am thinking these CI failures _might_ be transient. Going to `rerun tests` just to make sure, as things seem to pass locally for me.

**Comment by brandon-b-miller** [comment]
rerun tests

**Comment by brandon-b-miller** [comment]
@gpucibot merge
